src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+            }
+
+            return highlightedList;
+        }
+
+        private static ImmutableArray<AsyncCompletionData.CompletionFilterWithState> GetUpdatedFilters(
+            List<ExtendedFilterResult> filteredList,
+            ImmutableArray<AsyncCompletionData.CompletionFilterWithState> filters)
+        {
+            // See which filters might be enabled based on the typed code
+            var textFilteredFilters = filteredList.SelectMany(n => n.VSCompletionItem.Filters).Distinct();
+
+            // When no items are available for a given filter, it becomes unavailable
+            return ImmutableArray.CreateRange(filters.Select(n => n.WithAvailability(textFilteredFilters.Contains(n.Filter))));
+        }
+
+        private void MakeMostRecentItem(string item)
+        {
+            var updated = false;
+
+            while (!updated)
+            {
+                var oldItems = _recentItems;
+                var newItems = oldItems.Remove(item);
+
+                if (newItems.Length == MaxMRUSize)
+                {
+                    // Remove the least recent item.
+                    newItems = newItems.RemoveAt(0);
+                }
+
+                newItems = newItems.Add(item);
+                updated = ImmutableInterlocked.InterlockedCompareExchange(ref _recentItems, newItems, oldItems) == oldItems;
+            }
+        }
+
+        private static bool ShouldBeFilteredOutOfCompletionList(VSCompletionItem item, ImmutableArray<AsyncCompletionData.CompletionFilter> activeFilters)
+        {
+            foreach (var itemFilter in item.Filters)
+            {
+                if (activeFilters.Contains(itemFilter))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private void ItemCommitted(object sender, AsyncCompletionData.CompletionItemEventArgs e)
+        {
+            MakeMostRecentItem(e.Item.DisplayText);
+        }
+
+        private void SessionDismissed(object sender, EventArgs e)
+        {
+            if (sender is IAsyncCompletionSession session)
+            {
+                UnsubscribeEvents(session);
+            }
+        }
+
+        private void SubscribeEvents(IAsyncCompletionSession session)
+        {
+            UnsubscribeEvents(session);
+            session.ItemCommitted += ItemCommitted;
+
+            // We can be called multiple times for Update and for Sort during the lifecycle of the session.
+            // Then, the session can be dismissed out of our code. We should be sure, we do not keep references to dismissed sessions.
+            session.Dismissed += SessionDismissed;
+        }
+
+        private void UnsubscribeEvents(IAsyncCompletionSession session)
+        {
+            session.ItemCommitted -= ItemCommitted;
+            session.Dismissed -= SessionDismissed;
```

Comment:
what are the threading implications of all these calls?  can we put asserts in for how we expect to eb called?  is it always hte UI thread?  if so, assert it.  if not, we need locking.

Reply:
It is under background thread and should be locked. Added a singleton for this. Thank you!

---
In reply to: [232510768](https://github.com/dotnet/roslyn/pull/29016#discussion_r232510768) [](ancestors = 232510768)